language/vm/vm-runtime/vm-runtime-types/src/values/values_impl.rs

```diff
@@ -0,0 +1,2223 @@
+ *   internal invariants are violated.
+ *
+ **************************************************************************************/
+
+/// A reference to a Move struct that allows you to take a reference to one of its fields.
+#[derive(Debug)]
+pub struct StructRef(ContainerRef);
+
+/// A generic Move reference that offers two functinalities: read_ref & write_ref.
+#[derive(Debug)]
+pub struct Reference(ReferenceImpl);
+
+/// A Move value -- a wrapper around `ValueImpl` which can be created only through valid
+/// means.
+#[derive(Debug)]
+pub struct Value(ValueImpl);
+
+/// The locals for a function frame. It allows values to be read, written or taken
+/// reference from.
+#[derive(Debug)]
+pub struct Locals(Rc<RefCell<Container>>);
+
+/// An integer value in Move.
+#[derive(Debug)]
+pub enum IntegerValue {
+    U8(u8),
+    U64(u64),
+    U128(u128),
+}
+
+/// A Move struct.
+#[derive(Debug)]
+pub struct Struct(Container);
+
+/// A special value that lives in the global storage which you are allowed to take global
+/// references from. A global value also contains an internal flag, indicating whether
+/// the value has potentially been modified or not.
+#[derive(Debug)]
+pub struct GlobalValue {
+    status: Rc<RefCell<GlobalDataStatus>>,
+    container: Rc<RefCell<Container>>,
+}
+
+/***************************************************************************************
+ *
+ * Misc
+ *
+ *   Miscellaneous helper functions.
+ *
+ **************************************************************************************/
+impl Container {
+    fn len(&self) -> usize {
+        use Container::*;
+
+        match self {
+            General(v) => v.len(),
+            U8(v) => v.len(),
+            U64(v) => v.len(),
+            U128(v) => v.len(),
+            Bool(v) => v.len(),
+        }
+    }
+}
+
+impl ValueImpl {
+    fn new_container(container: Container) -> Self {
+        Self::Container(Rc::new(RefCell::new(container)))
+    }
+}
+
+impl Value {
+    pub fn is_valid_script_arg(&self, sig: &SignatureToken) -> bool {
+        match (sig, &self.0) {
+            (SignatureToken::U8, ValueImpl::U8(_)) => true,
+            (SignatureToken::U64, ValueImpl::U64(_)) => true,
+            (SignatureToken::U128, ValueImpl::U128(_)) => true,
+            (SignatureToken::Bool, ValueImpl::Bool(_)) => true,
+            (SignatureToken::Address, ValueImpl::Address(_)) => true,
+            (SignatureToken::ByteArray, ValueImpl::ByteArray(_)) => true,
+            _ => false,
+        }
+    }
+}
+
+/***************************************************************************************
+ *
+ * Borrows (Internal)
+ *
+ *   Helper functions to handle Rust borrows. When borrowing from a RefCell, we want
+ *   to return an error instead of panicking.
+ *
+ **************************************************************************************/
+fn try_borrow<T>(cell: &RefCell<T>) -> VMResult<Ref<T>> {
```

Comment:
Ah I see your confusion. I agree with it's not a good idea to pass `Ref` & `RefCell` across function boundaries in our current model (and to reassure we don't do this currently). 

It was never my intention for these to be perceived as an extra layer of indirection. So `RefCell::try_borrow()` returns `Result<Ref<T>, BorrowError>` and we want to rewrite it as `Result<Ref<T>, VMStatus>` (a.k.a `VMResult<Ref<T>>`) so we can propagate the error, and that's the only reason I added them. Otherwise you would have to call `map_err(|_| VMStatus::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
            .with_message(format!("vm borrow violation: {:?}", e)))` every time you do a borrow.